Alternative tutorial on implementing samplers (#468)
TuringLang/docs: tutorials/docs-17-implementing-samplers/index.qmd

\end{split}
\end{equation*}

i.e. we accept the proposal $\tilde{x}$ with probability $\alpha$ and reject it, thus sticking with our current iterate, with probability $1 - \alpha$.

## What we need from a model: [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl)

There are a few things we need from the "target" / "model" / density that we want to sample from:

1.  We need access to log-density *evaluations* $\log \gamma(x)$ so we can compute the acceptance ratio involving $\log \bar{\gamma}(x, p)$.
2.  We need access to log-density *gradients* $\nabla \log \gamma(x)$ so we can compute the Leapfrog steps $L_{\epsilon}(x, p)$.
3.  We also need access to the "size" of the model so we can determine the size of $M$.

Luckily for us, there is a package called [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) which provides an interface for *exactly* this!

To demonstrate how one can implement the "[LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) interface"[^1] we will use a simple Gaussian model as an example:

```{julia}
using LogDensityProblems: LogDensityProblems;

# Let's define some type that represents the model.

> struct NormalLogDensity{M<:AbstractVector{<:Real}}

Reviewer: Might it make sense to call this something else, like `IsotropicNormalModel` or something? It took me a moment to understand that this represents just the Gaussian model for which we can compute log densities, not a log density calculation.

Author: Good idea:)

## Implementing MALA in [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl)

Now that we've established that a model implementing the [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) interface provides us with all the information we need from $\log \gamma(x)$, we can address the question: given an object that implements the [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) interface, how can we define a sampler for it?

We're going to do this by making our sampler a sub-type of `AbstractMCMC.AbstractSampler` in addition to implementing a few methods from [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl). Why? Because it gets us *a lot* of functionality for free, as we will see later.

Moreover, [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl) provides a very natural interface for MCMC algorithms.

First, we'll define our `MALA` type

```{julia}
using AbstractMCMC

struct MALA{T,A} <: AbstractMCMC.AbstractSampler
    "stepsize used in the leapfrog step"
    ϵ_init::T
    "covariance matrix used for the momentum"
    M_init::A
end
```

> Notice how we've added the suffix `_init` to both the stepsize and the covariance matrix. We've done this because a `AbstractMCMC.AbstractSampler` should be *immutable*. Of course there might be many scenarios where we want to allow something like the stepsize and / or the covariance matrix to vary between iterations, e.g. during the burn-in / adaptation phase of the sampling process we might want to adjust the parameters using statistics computed from these initial iterations. But information which can change between iterations <span class="underline">should not go in the sampler itself</span>! Instead, this information should go in the sampler *state*.

Reviewer: Might the underline be better replaced with `*`? I tried to hover over it thinking it might be a link.

We will use the following sampler state for our `MALA` sampler:

```{julia}
struct MALAState{A<:AbstractVector{<:Real}}
    "current position"
    x::A
end
```

This might seem overly redundant: we're defining a type `MALAState` and it only contains a simple vector of reals.
In this particular case we indeed could have dropped this and simply used a `AbstractVector{<:Real}` as our sampler state, but typically, as we will see later, one wants to include other quantities in the sampler state.
For example, if we also wanted to adapt the parameters of our `MALA`, e.g. alter the stepsize depending on acceptance rates, in which case we should also put `ϵ` in the state, but for now we'll keep things simple.

We currently have two things:

1.  A `AbstractMCMC.AbstractSampler` implementation called `MALA`.
2.  A state `MALAState` for our sampler `MALA`.

That means that we're ready to implement the only thing that really matters: `AbstractMCMC.step`.

`AbstractMCMC.step` defines the MCMC iteration of our `MALA` given the current `MALAState`. Specifically, the signature of the function is as follows:

```{julia}
#| eval: false
function AbstractMCMC.step(

>     # The RNG to ensure reprodicibility.

Reviewer: ```suggestion
    # The RNG to ensure reproducibility.
```

    rng::Random.AbstractRNG,
    # The model that defines our target.
    model::AbstractMCMC.AbstractModel,
    # The sampler for which we're taking a `step`.
    sampler::AbstractMCMC.AbstractSampler,
    # The current sampler `state`.
    state;
    # Additional keyword arguments that we may or may not need.
    kwargs...
)
```

Moreover, there is a specific `AbstractMCMC.AbstractModel` which is used to indicate that the model that is provided implements the [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) interface: `AbstractMCMC.LogDensityModel`.

Since, as we discussed earlier, in our case we're indeed going to work with types that support the [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) interface, we'll define `AbstractMCMC.step` for such a `AbstractMCMC.LogDensityModel`.

> Note that `AbstractMCMC.LogDensityModel` has no other purpose; it has a single field called `logdensity`, and it does nothing else. But by wrapping the model in `AbstractMCMC.LogDensityModel`, it allows samplers which wants to work with [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) to define their `AbstractMCMC.step` on this type without running into method ambiguities.

Reviewer: ```suggestion
Note that `AbstractMCMC.LogDensityModel` has no other purpose; it has a single field called `logdensity`, and it does nothing else. But by wrapping the model in `AbstractMCMC.LogDensityModel`, it allows samplers that want to work with [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) to define their `AbstractMCMC.step` on this type without running into method ambiguities.
```

    x = state.x
    # Sample the momentum.
    p_dist = MvNormal(zeros(LogDensityProblems.dimension(model)), M)
    p = rand(rng, p_dist)
    # Propose using a single leapfrog step.
    x̃, p̃ = leapfrog_step(model, x, p, ϵ, M)
    # Accept or reject proposal.
    logp = LogDensityProblems.logdensity(model, x) + logpdf(p_dist, p)
    logp̃ = LogDensityProblems.logdensity(model, x̃) + logpdf(p_dist, p̃)
    logα = logp̃ - logp
    state_new = if log(rand(rng)) < logα
        # Accept.
        MALAState(x̃)
    else
        # Reject.
        MALAState(x)
    end
    # Return the "sample" and the sampler state.
    return state_new.x, state_new
end
```

Fairly straight-forward.

Of course, we haven't defined the `leapfrog_step` method yet, so let's do that:

```{julia}
function leapfrog_step(model, x, p, ϵ, M)
    # Update momentum `p` using "position" `x`.
    ∇logγ_x = last(LogDensityProblems.logdensity_and_gradient(model, x))
    p1 = p + (ϵ / 2) .* ∇logγ_x
    # Update the "position" `x` using momentum `p1`.
    x̃ = x + ϵ .* (M \ p1)
    # Update momentum `p1` using position `x̃`
    ∇logγ_x̃ = last(LogDensityProblems.logdensity_and_gradient(model, x̃))

>     p2 = p + (ϵ / 2) .* ∇logγ_x̃

Reviewer: Should this be 
```suggestion
    p2 = p1 + (ϵ / 2) .* ∇logγ_x̃
```
? I think that's what it is in the equations at the start of the page. I know nothing about these sampling algorithms though, so correct me if I'm wrong.

    # Flip momentum `p2`.
    p̃ = -p2

>     # Return.

Reviewer: ```suggestion
```
The return statement seems self-documenting.

samples_matrix = stack(samples)
```

```{julia}
# Compute the marginal means and standard deviations.
hcat(mean(samples_matrix; dims=2), std(samples_matrix; dims=2))
```

Let's visualize the samples

```{julia}
using StatsPlots
plot(transpose(samples_matrix[:, 1:10:end]), alpha=0.5, legend=false)
```

Look at that! Things are working; amazin'.

We can also exploit [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl)'s parallel sampling capabilities:

```{julia}
# Run separate 4 chains for 10 000 iterations using threads to parallelize.
num_chains = 4
samples = sample(
    model_with_grad,
    sampler,
    MCMCThreads(),
    10_000,
    num_chains;
    # Note we need to provide an initial state for every chain.
    initial_state=fill(state, num_chains),
    progress=false
)
samples_array = stack(map(stack, samples))
```

But the fact that we have to provide the `AbstractMCMC.sample` call, etc. with an `initial_state` to get started is a bit annoying. We can avoid this by also defining a `AbstractMCMC.step` *without* the `state` argument:

```{julia}
function AbstractMCMC.step(
    rng::Random.AbstractRNG,
    model_wrapper::AbstractMCMC.LogDensityModel,

>     sampler::MALA;

Reviewer: ```suggestion
    ::MALA;
```

    # NOTE: No state provided!
    kwargs...
)
    model = model_wrapper.logdensity
    # Let's just create the initial state by sampling using  a Gaussian.

>     x = randn(rng, LogDensityProblems.dimension(model))

Reviewer: This is only valid for our particular model, right? In that this may e.g. give negative values, which might not be valid parameters for some models. Don't want to overcomplicate the example, but might it be worth mentioning that in general one should e.g. sample from the prior?

EDIT: Noticed that you address this later. Maybe when you address it mention that the fix would be to change this method of `step`?

Author: Even if you use a different sample method here which samples values inside the support of the distribution, you'd still run into the same bug (just further down the callstack) because a gradient-based sampler like MALA will inevitably end up exploring the space outside of the support (unless you have a _tiny_ stepsize). So we can't really implement this sampler for models (in a generic fashion, without a lot of gluecode) in such a way that it also works with constrained distributions.

Heeeence I just avoid the issue entirely, demosntrate a case later and how we can use Turing.jl to "address" this.

    return x, MALAState(x)
end
```

Equipped with this, we no longer need to provide the `initial_state` everywhere:

```{julia}
samples = sample(model_with_grad, sampler, 10_000; progress=false)
samples_matrix = stack(samples)
hcat(mean(samples_matrix; dims=2), std(samples_matrix; dims=2))
```

## Using our sampler with Turing.jl

As we promised, all of this hassle of implementing our `MALA` sampler in a way that uses [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) and [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl) gets us something more than *just* an "automatic" implementation of `AbstractMCMC.sample`.

It also enables use with Turing.jl through the `externalsampler`, but we need to do one final thing first: we need to tell Turing.jl how to extract a vector of parameters from the "sample" returned in our implementation of `AbstractMCMC.step`. In our case, the "sample" is just a `Vector{<:Real}`, so we just need the following line:

```{julia}
# Load Turing.jl.
using Turing

# Overload the `getparams` method for our "sample" type, which is just a vector.

> Turing.Inference.getparams(::Turing.Model, x::AbstractVector{<:Real}) = x

Reviewer: I'm a bit confused by this, which looks like type piracy to me. Should this rather be defined for `MALAState`?

EDIT: Also just noticed that the docstring of `getparams` says "Return a named tuple of parameters", which adds to my confusion.

```

And with that, we're good to go!

```{julia}
# Our previous model defined as a Turing.jl model.
@model mvnormal_model() = x ~ MvNormal([-5., 0., 5.], I)
# Instantiate our model.
turing_model = mvnormal_model()
# Call `sample` but now we're passing in a Turing.jl `model` and wrapping
# our `MALA` sampler in the `externalsampler` to tell Turing.jl that the sampler
# expects something that implements LogDensityProblems.jl.
chain = sample(turing_model, externalsampler(sampler), 10_000; progress=false)
```

Pretty neat, eh?

### Models with constrained parameters

> One thing we've sort of glossed over in all of the above is that MALA, at least how we've implemented, requires $x$ to live in $\mathbb{R}^d$ for some $d > 0$. If some of the parameters were in fact constrained, e.g. we were working with a `Beta` distribution which has support on the interval $(0, 1)$, *not* on $\mathbb{R}^d$, we could easily end up outside of the valid range $(0, 1)$.

Reviewer: ```suggestion
One thing we've sort of glossed over in all of the above is that MALA, at least how we've implemented it, requires $x$ to live in $\mathbb{R}^d$ for some $d > 0$. If some of the parameters were in fact constrained, e.g. we were working with a `Beta` distribution which has support on the interval $(0, 1)$, *not* on $\mathbb{R}^d$, we could easily end up outside of the valid range $(0, 1)$.
```

```{julia}
chain = sample(turing_model, externalsampler(sampler; unconstrained=false), 10_000; progress=false)
```

The fun thing is that this still sort of works because

```{julia}
logpdf(Beta(3, 3), 10.0)
```

and so the samples that fall outside of the range are always rejected. But do notice how much worse all the diagnostics are, e.g. `ess_tail` is very poor compared to when we use `unconstrained=true`. Moreover, in more complex cases this won't just result in a "nice" `-Inf` log-density value, but instead will error:

```{julia}
@model function demo()
    σ² ~ truncated(Normal(), lower=0)
    # If we end up with negative values for `σ²`, the `Normal` will error.
    x ~ Normal(0, σ²)
end
sample(demo(), externalsampler(sampler; unconstrained=false), 10_000; progress=false)
```

As expected, we run into a `DomainError` at some point, while if we set `unconstrained=true`, letting Turing.jl transform the model to a unconstrained form behind the scenes, everything works as expected:

```{julia}
sample(demo(), externalsampler(sampler; unconstrained=true), 10_000; progress=false)
```

Neat!

> Similarly, which automatic differentiation backend one should use can be specified through the `adtype` keyword argument too. For example, if we want to use [ReverseDiff.jl](https://github.com/JuliaDiff/ReverseDiff.jl) instead of the default [ReverseDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl):

Reviewer: ```suggestion
Similarly, which automatic differentiation backend one should use can be specified through the `adtype` keyword argument too. For example, if we want to use [ReverseDiff.jl](https://github.com/JuliaDiff/ReverseDiff.jl) instead of the default [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl):
```

```{julia}
using ReverseDiff: ReverseDiff
# Specify that we want to use `AutoReverseDiff`.
sample(
    demo(),
    externalsampler(sampler; unconstrained=true, adtype=AutoReverseDiff()),
    10_000;
    progress=false
)
```

Double-neat.

## Summary

At this point it's worth maybe reminding ourselves what we did and also *why* we did it:

1.  We define our models in the [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) interface because it makes the sampler agnostic to how the underlying model is implemented.
2.  We implement our sampler in the [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl) interface, which just means that our sampler is a subtype of `AbstractMCMC.AbstractSampler` and we implement the MCMC transition in `AbstractMCMC.step`.

> 3.  (1) and (2) makes it so our sampler can be used with a wide range of model implementations, amongst them being models implemented in both Turing.jl and Stan. This gives you, the inference implementer, a large collection of models to test your inference method on, in addition to allowing users of Turing.jl and Stan to try out your inference method with minimal effort.

Reviewer: Quarto renders the `(1) and (2)` as "1. and (2)" with a funny indent. I think technically the `(1)` starts another nested enumerated list.
